### PR TITLE
support Ford Bronco Sport 2021-22, Maverick 2022

### PR DESCRIPTION
**Pull request by incognitojam**
- [x] new models
  - Bronco Sport 2021+
  - Maverick 2022+
- [x] test resume from stop works properly
- [x] check that TJA is toggled off
- [x] radar parser disabled because it's buggy
- [x] make steering good (ish)
  - https://github.com/commaai/cereal/pull/381
  
Note: In `LatCtlCurv_No_Actl`, positive curvature means left

### CODE REVIEW

**Comment by sshane** [comment]
Not sure if you were already going to, but feel free to split some of the refactors into smaller mergable PRs so you can reduce this diff

File: selfdrive/controls/lib/latcontrol_curvature.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+                               feedforward=desired_curvature,
+                               speed=CS.vEgo,
+                               freeze_integrator=freeze_integrator)
+
+      curvature_log.active = True
+      curvature_log.p = self.pid.p
+      curvature_log.i = self.pid.i
+      curvature_log.d = self.pid.d
+      curvature_log.f = self.pid.f
+      curvature_log.output = -output
```
Comment (incognitojam) [inline selfdrive/controls/lib/latcontrol_curvature.py:41]: need to calculate saturated?

File: selfdrive/car/ford/carcontroller.py [hunk 2]
```diff
@@ -46,33 +30,35 @@ def update(self, CC, CS):
...
-
     # send steering commands at 20Hz
     if (self.frame % CarControllerParams.LKAS_STEER_STEP) == 0:
-      # use LatCtlPath_An_Actl to actuate steering
-      path_angle = math.radians(apply_angle) / CarControllerParams.LKAS_STEER_RATIO
+      if CC.latActive:
+        lca_rq = 1
+
+        # use LatCtlCurv_No_Actl to actuate steering
+        # TODO: apply rate limits
```
Comment (incognitojam) [inline selfdrive/car/ford/carcontroller.py:51]: add rate limits back

**Comment by incognitojam** [comment]
Route on `ford-devel` which is based on this branch: `26b2cace68e36212|2022-11-05--20-51-49`

File: selfdrive/controls/lib/latcontrol_curvature.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+from selfdrive.controls.lib.latcontrol import LatControl, MIN_STEER_SPEED
+from selfdrive.controls.lib.pid import PIDController
+
+STEER_ANGLE_SATURATION_THRESHOLD = 2.5  # Degrees
+
+
+class LatControlCurvature(LatControl):
+  def __init__(self, CP, CI):
+    super().__init__(CP, CI)
+    self.steer_max = 0.02
```
Comment (incognitojam) [inline selfdrive/controls/lib/latcontrol_curvature.py:13]: this feels like a hack. I thought to use [-1.0, 1.0] and apply in carcontroller, but I think would have to test new PID coefficients?

**Comment by incognitojam** [comment]
another route on `ford-devel` (rebased on this branch):
`26b2cace68e36212|2022-11-21--23-21-12`

File: selfdrive/controls/lib/latcontrol_curvature.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
+import math
+
+from cereal import log
+from selfdrive.controls.lib.latcontrol import LatControl, MIN_STEER_SPEED
+from selfdrive.controls.lib.pid import PIDController
+
+STEER_ANGLE_SATURATION_THRESHOLD = 2.5  # Degrees
```
Comment (sshane) [inline selfdrive/controls/lib/latcontrol_curvature.py:7]: move to latcontrol.py and import

File: selfdrive/controls/lib/latcontrol_curvature.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+
+class LatControlCurvature(LatControl):
+  def __init__(self, CP, CI):
+    super().__init__(CP, CI)
+    self.pid = PIDController((CP.lateralTuning.pid.kpBP, CP.lateralTuning.pid.kpV),
+                             (CP.lateralTuning.pid.kiBP, CP.lateralTuning.pid.kiV),
+                             k_f=CP.lateralTuning.pid.kf, pos_limit=self.steer_max, neg_limit=-self.steer_max)
```
Comment (sshane) [inline selfdrive/controls/lib/latcontrol_curvature.py:15]:
```suggestion
                             k_f=CP.lateralTuning.pid.kf)
```

Let's leave it up to the PID defaults (practically unlimited). Any number seems arbitrary to me. 0.02 is 20 m/s^2 of lateral accel at 70 mph, and 1.0 (steer_max) is similar in that vein.

File: selfdrive/controls/controlsd.py [hunk 6]
```diff
@@ -608,9 +611,10 @@ def state_control(self, CS):
                                                                                        lat_plan.psis,
                                                                                        lat_plan.curvatures,
                                                                                        lat_plan.curvatureRates)
-      actuators.steer, actuators.steeringAngleDeg, lac_log = self.LaC.update(CC.latActive, CS, self.VM, lp,
-                                                                             self.last_actuators, self.steer_limited, self.desired_curvature,
-                                                                             self.desired_curvature_rate, self.sm['liveLocationKalman'])
+      actuators.steer, actuators.steeringAngleDeg, actuators.curvature, lac_log = \
+        self.LaC.update(CC.latActive, CS, self.VM, lp,
+                        self.last_actuators, self.steer_limited, self.desired_curvature,
+                        self.desired_curvature_rate, self.sm['liveLocationKalman'])
```
Comment (sshane) [inline selfdrive/controls/controlsd.py:617]:
```suggestion
      actuators.steer, actuators.steeringAngleDeg, actuators.curvature, lac_log = \
        self.LaC.update(CC.latActive, CS, self.VM, lp,
                        self.last_actuators, self.steer_limited, self.desired_curvature,
                        self.desired_curvature_rate, self.sm['liveLocationKalman'])
```

We might want to start passing in `actuators` and the lateral controller can modify it

File: selfdrive/controls/controlsd.py [hunk 7]
```diff
@@ -620,7 +627,7 @@ def state_control(self, CS):
         if CC.latActive:
           steer = clip(self.sm['testJoystick'].axes[1], -1, 1)
           # max angle is 45 for angle-based cars
-          actuators.steer, actuators.steeringAngleDeg = steer, steer * 45.
+          actuators.steer, actuators.steeringAngleDeg, actuators.curvature = steer, steer * 45., steer * 0.02
```
Comment (incognitojam) [inline selfdrive/controls/controlsd.py:630]: maybe we could do something smarter here

File: selfdrive/controls/lib/latcontrol_pid.py [hunk 8]
```diff
@@ -45,4 +45,5 @@ def update(self, active, CS, VM, params, last_actuators, steer_limited, desired_
       pid_log.output = output_steer
       pid_log.saturated = self._check_saturation(self.steer_max - abs(output_steer) < 1e-3, CS, steer_limited)
 
-    return output_steer, angle_steers_des, pid_log
```
Comment (incognitojam) [inline selfdrive/controls/lib/latcontrol_pid.py:48]: I didn't set `actuators.steeringAngleDeg` because we already set `pid_log.steeringAngleDesiredDeg`

File: selfdrive/controls/controlsd.py [hunk 9]
```diff
@@ -642,6 +649,8 @@ def state_control(self, CS):
         # TODO use desired vs actual curvature
         if self.CP.steerControlType == car.CarParams.SteerControlType.angle:
           steering_value = actuators.steeringAngleDeg
+        elif self.CP.steerControlType == car.CarParams.SteerControlType.curvature:
+          steering_value = actuators.curvature
```
Comment (incognitojam) [inline selfdrive/controls/controlsd.py:653]: User reported that "steering limit reached" alert never appears - this line is suspect but I tried it both `-ve` and `+ve` and it didn't appear. maybe the issue is somewhere else.

File: selfdrive/car/tests/routes.py [hunk 10]
```diff
@@ -16,8 +16,10 @@
 
 # TODO: add routes for these cars
 non_tested_cars = [
+  FORD.BRONCO_SPORT_MK1,
```
Comment (incognitojam) [inline selfdrive/car/tests/routes.py:19]: @adeebshihadeh We have some routes for the Bronco Sport, maybe we could use one of those?

File: selfdrive/controls/lib/latcontrol_pid.py (hunk 8 above)
Comment (incognitojam) [reply selfdrive/controls/lib/latcontrol_pid.py:48]: oh this might be why process replay changed

File: selfdrive/controls/lib/latcontrol_curvature.py [hunk 11]
```diff
@@ -0,0 +1,48 @@
+import math
+
+from cereal import log
+from selfdrive.controls.lib.latcontrol import MIN_STEER_SPEED, STEER_ANGLE_SATURATION_THRESHOLD, LatControl
+from selfdrive.controls.lib.pid import PIDController
+
+
+class LatControlCurvature(LatControl):
```
Comment (sshane) [inline selfdrive/controls/lib/latcontrol_curvature.py:8]: you can subclass `LatControlPID` and remove the init function here

File: selfdrive/car/ford/interface.py [hunk 12]
```diff
@@ -1,40 +1,52 @@
...
   def _get_params(ret, candidate, fingerprint, car_fw, experimental_long):
     ret.carName = "ford"
-    ret.dashcamOnly = True
     ret.safetyConfigs = [get_safety_config(CarParams.SafetyModel.ford)]
 
-    # Angle-based steering
-    ret.steerControlType = CarParams.SteerControlType.angle
-    ret.steerActuatorDelay = 0.4
-    ret.steerLimitTimer = 1.0
+    # These cars are dashcam only until steering safety is implemented
```
Comment (adeebshihadeh) [inline selfdrive/car/ford/interface.py:14]:
```suggestion
    # These cars are dashcam only until the port is finished
```

File: selfdrive/car/ford/fordcan.py [hunk 13]
```diff
@@ -4,161 +4,164 @@
...
-    elif hud_control.rightLaneVisible:
-      lines += 5
-  elif main_on:
-    lines = 0
-  else:
-    if hud_control.leftLaneDepart:
-      lines = 3  # WarnLeft_NoRight
-    elif hud_control.rightLaneDepart:
-      lines = 15  # NoLeft_WarnRight
+class FordCAN:
```
Comment (adeebshihadeh) [inline selfdrive/car/ford/fordcan.py:7]: any reason this is a class?

File: selfdrive/car/ford/carcontroller.py [hunk 14]
```diff
@@ -46,33 +31,33 @@ def update(self, CC, CS):
...
-      lca_rq = 0
-      apply_angle = 0.
-
     # send steering commands at 20Hz
     if (self.frame % CarControllerParams.LKAS_STEER_STEP) == 0:
-      # use LatCtlPath_An_Actl to actuate steering
-      path_angle = math.radians(apply_angle) / CarControllerParams.LKAS_STEER_RATIO
+      if CC.latActive:
+        # use LatCtlCurv_No_Actl to actuate steering
+        # IMPORTANT! left is positive
```
Comment (adeebshihadeh) [inline selfdrive/car/ford/carcontroller.py:50]: keep comments like this in one place. right now, some are here and others in fordcan.py

File: selfdrive/car/ford/fordcan.py (hunk 13 above)
Comment (incognitojam) [reply selfdrive/car/ford/fordcan.py:7]: just to remove packer arg, and then when we get around to CAN-FD we will need to specify a different bus too

Comment (adeebshihadeh) [reply selfdrive/car/ford/fordcan.py:7]: i'd rather not introduce new patterns here

Comment (incognitojam) [reply selfdrive/car/ford/fordcan.py:7]: oh I just saw that tesla already did it https://github.com/commaai/openpilot/blob/master/selfdrive/car/tesla/teslacan.py

Comment (incognitojam) [reply selfdrive/car/ford/fordcan.py:7]: I can revert for now

File: selfdrive/controls/lib/latcontrol_curvature.py [hunk 15]
```diff
@@ -1,21 +1,11 @@
...
+from selfdrive.controls.lib.latcontrol import MIN_STEER_SPEED, STEER_ANGLE_SATURATION_THRESHOLD
+from selfdrive.controls.lib.latcontrol_pid import LatControlPID
 
 
-class LatControlCurvature(LatControl):
-  def __init__(self, CP, CI):
-    super().__init__(CP, CI)
-    self.pid = PIDController((CP.lateralTuning.pid.kpBP, CP.lateralTuning.pid.kpV),
-                             (CP.lateralTuning.pid.kiBP, CP.lateralTuning.pid.kiV),
-                             k_f=CP.lateralTuning.pid.kf)
```
Comment (sshane) [inline selfdrive/controls/lib/latcontrol_curvature.py]: Oh I forgot now the PID limits (eg. `pos_limit`) are set in LatControlPID

File: selfdrive/controls/lib/latcontrol_curvature.py (hunk 11 above)
Comment (sshane) [reply selfdrive/controls/lib/latcontrol_curvature.py:8]: you can subclass again if you do something like https://github.com/commaai/openpilot/pull/26744/commits/d8ada72314ed2990527e3c13be1ae4cb1528b08f

File: selfdrive/car/ford/carstate.py [hunk 16]
```diff
@@ -14,7 +14,7 @@ def __init__(self, CP):
     super().__init__(CP)
     can_define = CANDefine(DBC[CP.carFingerprint]["pt"])
     if CP.transmissionType == TransmissionType.automatic:
-      self.shifter_values = can_define.dv["Gear_Shift_by_Wire_FD1"]["TrnGear_D_RqDrv"]
+      self.shifter_values = can_define.dv["Gear_Shift_by_Wire_FD1"]["TrnRng_D_RqGsm"]
```
Comment (sshane) [inline selfdrive/car/ford/carstate.py:17]: Can we do all this extra stuff in another PR? Trying to make this PR as small as possible

**Comment by sshane** [comment]
Split into these PRs:
- https://github.com/commaai/openpilot/pull/27090
- https://github.com/commaai/openpilot/pull/27092
- https://github.com/commaai/openpilot/pull/27094
- https://github.com/commaai/openpilot/pull/27093
- https://github.com/commaai/openpilot/pull/27091
- https://github.com/commaai/openpilot/pull/27096
- https://github.com/commaai/openpilot/pull/27095
